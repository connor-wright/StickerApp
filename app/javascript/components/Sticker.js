import React from "react";
import PropTypes from "prop-types";
class Sticker extends React.Component {
  constructor(props)
  {
    super(props);
    if(props.pos)
    {
      this.state = {
        url: props.url,
        xpos: props.pos.posX,
        ypos: props.pos.posY
      };
    }
    else{
      this.state = {
        url: props.url
      }
    }
  }
  
  render () {
    const {xpos, ypos, url} = this.state;
    const Stickerstyle = {
      top: ypos,
      left: xpos
    };
    
    return (
      <div style={Stickerstyle} className="sticker">
        <img src={url}/>
        <div>
          xpos: {xpos}
          ypos: {ypos}
        </div>
      </div>
    );
  }
}

export default Sticker;
