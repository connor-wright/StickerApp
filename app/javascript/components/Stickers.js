import React from "react"
import PropTypes from "prop-types"
class Stickers extends React.Component {
  render () {
    return (
      <React.Fragment>
      </React.Fragment>
    );
  }
}

export default Stickers
