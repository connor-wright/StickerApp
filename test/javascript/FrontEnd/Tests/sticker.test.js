import React from 'react';
import Sticker from 'Sticker';
import Stickers from 'Stickers';
import { mount } from 'enzyme';
import {runAllPromises} from '../Utils/utils';
const BackendAPI = require('BackendAPI');

jest.mock('BackendAPI');
const GetStickers = BackendAPI.GetStickers;
const GetImgurImg = BackendAPI.GetImgurImg;
const PostSticker = BackendAPI.PostSticker;
let stickersComponent = null;
let activeId = null;

//There are some weird behaviors with enzyme and to match snapshot
//they dont seem to be in sync. 
//TODO figure this out
describe('<Stickers />' , () => {
  beforeAll(() => {
    GetStickers.mockResolvedValue([]);
  });
  
  beforeEach(() => {
    activeId = 'Tacos';
    stickersComponent = mount(<Stickers activeId={activeId}/>);
  });
  
  afterEach(() => {
    stickersComponent.unmount();
  });
  
  it('Should request stickers on mount', () => {
    expect(GetStickers).toHaveBeenCalled();
  });
  
  it('Should display stickers after mount', async () => {
    //arrange
    GetStickers.mockResolvedValue([{
      url: 'https://imgur.com/gallery/3nSMulw', 
      xpos: '25',
      ypos: '45',
      id: 'Tacos'
    }]);
    
    //act
    stickersComponent.unmount();
    stickersComponent = mount(<Stickers/>);
    await runAllPromises();
    stickersComponent.update();
    
    //assert
    expect(stickersComponent).toMatchSnapshot();
  });
  
  it('Should add a sticker based on active ID on click', async () => {
    //arrange
    //weird issue where react will think there is two sticker 
    //elements with the same keys even though there isnt. 
    activeId = 'TwoTacos';
    stickersComponent.unmount();
    stickersComponent = mount(<Stickers activeId={activeId}/>);
    await runAllPromises();
    
    const expectedXpos = 25;
    const expectedYpos = 75;
    const url = 'https://imgur.com/gallery/3nSMulw';
    const expectedPostNewPhotoArgs = {
      img_id: activeId,
      url: url,
      xpos: expectedXpos,
      ypos: expectedYpos
    };
    
    GetImgurImg.mockResolvedValue({data: {
      id: activeId,
      link: url
    }});
    PostSticker.mockResolvedValue({
      url: url, 
      xpos: expectedXpos,
      ypos: expectedYpos,
      id: activeId
    });
    
    //act
    stickersComponent.simulate(
      'click', {clientX: expectedXpos, clientY: expectedYpos});
    await runAllPromises();
    await stickersComponent.update();
    
    //assert
    expect(GetImgurImg.mock.calls).toContainEqual([activeId]);
    expect(PostSticker.mock.calls).toContainEqual([expectedPostNewPhotoArgs]);
    //TODO need to assert that the sticker is created
  });
});