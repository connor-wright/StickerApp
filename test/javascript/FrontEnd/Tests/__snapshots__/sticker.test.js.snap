// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Stickers /> Should display stickers after mount 1`] = `
<Stickers>
  <div
    className="Stickers"
    onClick={[Function]}
  >
    <Sticker
      key="Tacos"
      url="https://imgur.com/gallery/3nSMulw"
      xpos="25"
      ypos="45"
    />
  </div>
</Stickers>
`;
